pkg/apis/core/validation/validation.go

```diff
@@ -6094,3 +6145,177 @@ func ValidateSpreadConstraintNotRepeat(fldPath *field.Path, constraint core.Topo
+		allErrs = append(allErrs, field.Invalid(specPath.Child("clusterIPs"), service.Spec.ClusterIPs, "maximum allowed entries in clusterIPs is two"))
+	}
+
+	// dual stack feature gate validation
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) && len(service.Spec.ClusterIPs) > 1 {
+		allErrs = append(allErrs, field.Invalid(clusterIPsField, service.Spec.ClusterIPs, "multiple ClusterIPs are only allowed when IPv6DualStack feature gate is enabled"))
+	}
+
+	// at this stage if there is an invalid ip or misplaced none/empty string
+	// it will skew the error messages (bad index || dualstackness of already bad ios). so we
+	// stop here if there are errors in clusterIPs validation
+	if hasInvalidIPs || hasMisplacedNoneEmptyString {
+		return allErrs
+	}
+
+	// must be dual stacked ips if they are more than one ip
+	if len(service.Spec.ClusterIPs) > 1 /*meaning: it does not have a None or empty string */ {
+		dualStack, err := netutils.IsDualStackIPStrings(service.Spec.ClusterIPs)
+		if err != nil { // though we check for that earlier. safe > sorry
+			allErrs = append(allErrs, field.InternalError(clusterIPsField, fmt.Errorf("failed to check for dual stack with error:%v", err)))
+		}
+
+		// We only support one from each IP family (i.e. max two IPs in this list).
+		if !dualStack {
+			allErrs = append(allErrs, field.Invalid(clusterIPsField, service.Spec.ClusterIPs, "may specify no more than one IP for each IP family"))
+		}
+	}
+
+	// field dependency
+	// if families > 1 then .spec.ipFamilyPolicy must not be set to SingleStack
+	if len(service.Spec.IPFamilies) > 1 &&
+		service.Spec.IPFamilyPolicy != nil &&
+		*(service.Spec.IPFamilyPolicy) == core.SingleStack {
+		allErrs = append(allErrs, field.Invalid(ipFamilyPolicyField, service.Spec.IPFamilyPolicy, "must be set to RequireDualStack or PreferDualStack when multiple ip families are provided"))
+	}
+
+	// if clusterIPs > 1 then .spec.ipFamilyPolicy must be set to RequireDualStackl
+	if len(service.Spec.ClusterIPs) > 1 &&
+		service.Spec.IPFamilyPolicy != nil &&
+		*(service.Spec.IPFamilyPolicy) == core.SingleStack {
+		allErrs = append(allErrs, field.Invalid(ipFamilyPolicyField, service.Spec.IPFamilyPolicy, "must be set to RequireDualStack or PreferDualStack when multiple cluster ips are provided"))
+	}
+
+	// match clusterIPs to their families, if they were provided
+	// ignore when None and empty string are used
+	if len(service.Spec.ClusterIPs) > 0 && service.Spec.ClusterIPs[0] != core.ClusterIPNone && service.Spec.ClusterIPs[0] != "" && len(service.Spec.IPFamilies) > 0 {
+		for i, ip := range service.Spec.ClusterIPs {
+			if i > (len(service.Spec.IPFamilies) - 1) {
+				break // no more families to check
+			}
+
+			// 4=>6
+			if service.Spec.IPFamilies[i] == core.IPv4Protocol && netutils.IsIPv6String(ip) {
+				allErrs = append(allErrs, field.Invalid(clusterIPsField.Index(i), service.Spec.ClusterIPs, fmt.Sprintf("expected IPv4 as set indicated by ipFamilies[%v]", i)))
```

Comment:
I am not sure we need distinct errors, but OK :)

Wording nit: "expected IPv4 as set indicated by" => ""expected IPv4 as indicated by"  (drop "set")